### ISSUE TITLE
Constructor parameters could not be defined on a different level of inheritance hierarchy

### ISSUE THREAD

**Issue by digitalbuddha** [immutables/immutables#292]
Context:
I've recently started using https://github.com/square/sqldelight with Immutables to allow me to dump my immutable models into a sqllite db.  
The lib takes dsl file like: 
https://gist.github.com/digitalbuddha/588c5aeb82349a29230bf36d60b1eb42

and generates a java interface:
https://gist.github.com/digitalbuddha/d47d69d37f82a3afc9c76609abf4a282

The library authors recommend using the above interface with autovalue like so:

```
@AutoValue
public abstract class Payload implements PayloadModel {
  public static final Mapper<Payload> MAPPER = new Mapper<>(AutoValue_Payload::new);

  public static final class Marshal extends PayloadMarshal<Marshal> { }
}
```

Being the avid Immutables fan that I am I tried doing something similar with immutables but got following error:

Constructor parameters could not be defined on a different level of inheritance hierarchy,  generated constructor API would be unstable. To resolve, you can redeclare (override) each inherited constuctor parameter in this abstract value type.

The best I can do with Immutables is something like
https://gist.github.com/digitalbuddha/f4d9a4842e0266ba1a57c106d562b7bd

Could you explain to me why different levels of constructor parameters is bad and if there is another way I can achieve the same solution without the code duplication.  

Thank you kindly.

**Comment by elucash** [comment]
There are 2 things here. I'll explain you why we have such an error, but here, I unable to reproduce it as it works for me.

```java
@Value.Immutable
@Value.Style(allParameters = true)
public abstract class Payload implements PayloadModel {
	public static final PayloadModel.Mapper<Payload> MAPPER = new PayloadModel.Mapper<>(
			ImmutablePayload::of);
}
```
Can you create some simple repro case with a build-file? I'm attaching mine
[repro.zip](https://github.com/immutables/immutables/files/199084/repro.zip)

Returning to the actual error, it was introduced to prohibit unwieldy inheritance of constructor parameters.

```java
interface X {
   @Parameter int x();
}
interface Y {
    @Parameter int y();
}
interface Z extends X, Y {
    @Parameter int z();
}
@Immutable
interface Point extends  X, Y, Z {}
```

The problem here what the constructor would be generated? `of(X, Y, Z)` or `of(Z, Y, X)` or whatever? What will happen if you add additional parameter to one of the supertype then?
To cope with the low predictability of not alway apparent ordering and number of parameters, we allow inheritance of parameters, but require so that all of them would be defined in one supertype or abstract value type, allowing abstract value type to actually override all of parameters in case of conflicts between supertypes. It is perfectly ok to inherit other non-parameter non-mandatory attributes scattered accross multiple supertypes.

Please, let us know the details of compiler/env you use, and/or try to isolate to the minimum repro case.

**Comment by digitalbuddha** [comment]
Hi @elucash you are correct this works:
```
@Value.Immutable
@Value.Style(allParameters = true)
public abstract class Payload implements PayloadModel {
    public static final PayloadModel.Mapper<Payload> MAPPER = new PayloadModel.Mapper<>(
            ImmutablePayload::of);
}
```

The problem is more nuancey than I thought.  While the above works, what won't work is something like this:

```
@Value.Immutable
@Value.Style(allParameters = true)
public abstract class Appointment implements AppointmentModel {

    public static final AppointmentModel.Mapper<Appointment> MAPPER = new AppointmentModel.Mapper<>(
            ImmutableAppointment::of);
    
    public abstract List<Payload> payloads();

}
```
Where `AppointmentModel` contains a bunch of fields and `payloads` is a list of some value object that is not in the base `AppointmentModel`


Sorry if this is confusing but here's what I want to do: I'd like to use the above `Appointment` class to get data from some API.  I then want to save the Appointment into a db.  Since I don't have flat data I'd instead lazy save/load the List of Payloads to the db separately. Ideally I'd like to be able to have a class that looks like the `Appointment` class but works like:
public abstract class Appointment implements AppointmentModel {

 ```
@Value.Immutable
@Value.Style(allParameters = true)
   public static final AppointmentModel.Mapper<Appointment> MAPPER = new AppointmentModel.Mapper<>(
            ImmutableAppointment::of);
    
    @Nullable
    public Long _id() {
        return 0;
    }

    @Value.Parameter
    public abstract long id();

    @Value.Parameter
    @Nullable
    public abstract Integer company_id();

    @Value.Parameter
    @Nullable
    public abstract Integer product_id();

 
    public abstract List<Payload> payloads();  
    }

One more way to say it: I think I need a @Value.Style(allNonCollectionParameters = true)

Thanks again.

**Comment by elucash** [comment]
I think its better to cancel out parameters with `@Value.Parameter(false)` or so on per-attribute level, `allNonCollectionParameters` is too ad hoc for a general case. Of course, there is a tension to just suggest annotates parameters directly instead of using `allParameters = true`, but at the same time the sqldelight shows how useful it is to have `allParameters`, so now we need to cancel out parameters.

**Comment by digitalbuddha** [comment]
@Value.Parameter(false) would be absolutely perfect and at the necessary granularity.  Tag me on PR so I can see how to make such an enhancement, I'd love to start figuring out how I can contribute more to this project.

**Comment by elucash** [comment]
It was quite simple to implement, but a bit too esoteric to ask for PR there. I'll add some additional bug-fixes and then release. In a meantime you can try snapshot version.

**Comment by digitalbuddha** [comment]
compile my own or is there a way I can pull snapshot from maven?

**Comment by elucash** [comment]
both ways, you can compile and install locally or use

```xml
  <!--
    this repository allows to quickly evaluate unreleased upstream components
    from sonatype oss repository -->
  <repositories>
    <repository>
      <id>sonatype-nexus-snapshots</id>
      <name>Sonatype Nexus Snapshots</name>
      <url>https://oss.sonatype.org/content/repositories/snapshots</url>
      <releases>
        <enabled>false</enabled>
      </releases>
      <snapshots>
        <enabled>true</enabled>
      </snapshots>
    </repository>
  </repositories>
```
(or Gradle variation of the same)

**Comment by digitalbuddha** [comment]
all set thank you for the handle holding :100:

**Comment by digitalbuddha** [comment]
Top of the morning!  Snapshot went bad on me this morning. Is there a different way I can pull the latest version.  Thank you kindly.

**Comment by digitalbuddha** [comment]
Nevermind, sorry seems like it was just a blip for a second, compiling fine again

**Comment by elucash** [comment]
Released in v 2.1.16 Thank you!

**Comment by renedewaele** [comment]
Hi,

Really loving the framework. Thanks for all your work!

I've just run into this "Constructor parameters could not be defined on a different level..." issue in a project now. It forces `@override`'s in the subclass which is ugly boilerplate this framework is usually so good to remove.

I'm still not sure why the order of constructor parameters would be unpredictable. Would it not be possible to order by declared inheritance order first and inheritance level second? I.e. your example:

```
interface X {
   @Parameter int x();
}

interface Y {
    @Parameter int y();
}

interface Z extends X, Y {
    @Parameter int z();
}

@Immutable
interface Point extends  X, Y, Z {}
```

Would end up in constructor `ImmutablePoint.of(x, y, z)` because X was declared before Y was declared before Z. If the example is tweaked to:

```
interface X {
   @Parameter int x();
}

interface Y {
    @Parameter int y();
}

interface Z extends X, Y {
    @Parameter int z();
}

@Immutable
interface Point extends  Y, Z {}
```

the constructor becomes: `ImmutablePoint.of(y, x, z)`.

The suggested rule is arbitrary of course but the result should be predictable, right? What am I overlooking?

Cheers,
Rene

**Comment by elucash** [comment]
Hi @renedewaele There are 2 points why I think the limitation still make sense:

1) While you can assume the ordering of the constructor parameters initially, subsequent addition of attributes to intermediate interfaces results in the insertion of parameters in lot less predictable order. This is complicated by the fact the same attribute can be defined by more that one super-interface

```java
interface X {
   @Parameter int x();
   @Parameter int u(); // <-- later addition
}
interface Y {
    @Parameter int y();
}
interface Z extends X, Y {
    @Parameter int z();
}
// how constructor will look now?
// or 
interface X {
   @Parameter int x();
   @Parameter int z(); // <-- how about now?
}
interface Y {
    @Parameter int y();
}
interface Z extends X, Y {
    @Parameter int z();
}
```

2) Compilation/annotation processing does not always operate on the source code only. If you have supertypes coming from external libraries or during incremental compilation, compiles use class files to read supertypes. As it stands, the exact ordering of the supertypes and their methods is not portable across compilers. Maybe they use hash-sets with object identity hashcodes for elements or something else, but we've observed real problems with builds where the order of parameters was changing from time to time, making builds not reproducible.

**Comment by renedewaele** [comment]
Hi @elucash,

Thanks for your clarification. 

Regarding your first argument, I still don't see that big of an issue. In the first case the constructor becomes `ImmutablePoint.of(x, u, y, z)` and in the second case it would be `ImmutablePoint.of(x, z, y)`. Again this is all based on the initially proposed ordering rules which were pretty arbitrary but do yield predictable results.

If your point is that users may accidentally invoke the constructor the wrong way, as in what if `ImmutablePoint.of(x, z, y)` becomes `ImmutablePoint.of(x, y, z)` after a change in the signature of X, Y or Z, I'd say it is still worth it. First, that's a pretty rare situation and it is up to the user to verify that everything still works ok (check places that invoke the constructor). In case the user did not check there's probably still a unit test here or there that will fall over.

Right now extending interfaces with constructor parameters introduces boilerplate for everyone. I'd say that's a worse problem.

Regarding your second point. I get that things get more complicated when extending classes from external libraries so a warning may be in order in the documentation and perhaps even during execution of the annotation processor. However, I'd bet that extending an `@Value.Immutable` annotated type from a library is another pretty rare situation and that the framework should allow this.

But again, great framework! Keep up the good work.

**Comment by elucash** [comment]
I would definitely agree that the first issue may be ignored with a bit or resolve and discipline (and having right expectations). The second issue is not about very rare situations. Every incremental build is risking to change the order, moreover, consecutive full builds may give a different order. Possibly, you can get around it by chance or by only using full builds with all supertypes available as sources. It is also depends on compiler, i.e. Javac vs ECJ + version variations for each compiler.
If we disable this check, the next week we will get new bug report about unstable parameter order, and we will be unable to fix it.
As a matter or experiment, we can turn it into a warning when using with `allParameters` style. Such a warning can then be [suppressed](http://immutables.github.io/immutable.html#warnings)
@renedewaele  WDYT about warning instead of error?

**Comment by renedewaele** [comment]
@elucash I've never used the `allParameters` style. Does that mean that all `Optional` and `@Nullable` attributes also need to be explicitly declared each time the object is built? If so, that would not help me as I'm using it for objects with lots of unknowns.

Would it instead be possible to explicitly enable the functionality without giving an error but with only a warning? The current behavior (error message) would remain the default but if you're using the right compiler etc you would be able to override the default settings at your own risk.

**Comment by elucash** [comment]
We'll go with the warning always instead of the error in the next release (2.3.10 or 2.4). We'll see how it goes.

**Comment by elucash** [comment]
released in 2.4.0, thank you!
